### DISCUSSION TITLE
How to generate exe?

### DISCUSSION THREAD

**Discussion by meoowe** [korlibs/korge#2037] [Q&A] [answered]
Hello, how can I generate a Windows executable? Do I make a JvmJar and then use a tool to generate an exe, or is there another way that I am not aware of?

**Comment by maik-mursall** [comment, marked as answer]
Heya @meoowe 👋🏻 

You can use the Gradle Command `packageJvmWindowsApp` :)

Here are the commands for the following targets: (JVM & JS)
![image](https://github.com/korlibs/korge/assets/54634184/9a3da98a-61fa-4bf8-9a25-8384df38cf40)

You'll find the built project in `<your-module>/build/platforms`.

Below you'll see it within an multi-module project (client, server, shared):
![image](https://github.com/korlibs/korge/assets/54634184/8c698c09-43da-4e43-ba63-0ceab1871e30)

Be aware, that you need to ship the entire folder (`jvm-win32` in this case) for it to work :)
(You can just rename it tho :))

**Reply by meoowe** [reply]
Hmm. I can't see the `packageJvmWindowsApp`.
![Screenshot 2023-11-24 161105](https://github.com/korlibs/korge/assets/96916064/e970a0ce-708d-4a6b-94fb-6292fea89594)

**Reply by maik-mursall** [reply]
Are you on v5.1.0? Since this feature is only available since then 😅.

If not try changing it to "5.1.0" in your `gradle/libs.versions.toml`-File.

Make sure to re-sync the Gradle project again then :)

**Reply by meoowe** [reply]
THANK YOU!!  You've helped me so much!

**Reply by maik-mursall** [reply]
Glad to help :D 🎉

**Reply by meoowe** [reply]
In an error, it says that a parameter( `this` ) is null ???

**Reply by maik-mursall** [reply]
Where does the error come from? During building or running the app? :)

It would also help a lot, if you could post the error message 😅

**Reply by meoowe** [reply]
Error message - `Parameter specified as non-null is null: method korlibs.korge.gradle.targets.jvm.JvmKt.getMainJvmCompilation, parameter <this>`

this happens during the `packageJvmFatJar` task while building.
sorry for disturbing you again

**Reply by maik-mursall** [reply]
I just found out the cause of your error :D

In your `gradle.properties` file, delete the last two lines like following:
![image](https://github.com/korlibs/korge/assets/54634184/887b497b-e674-490b-b914-1de15949aac5)

After another Gradle Re-Sync you should be able to build your app :)

Something within `packageJvmFatJar` seems to be incompatible with the Gradle Cache 😅.

I can reproduce this with the "Hello World" Example :).

**Reply by maik-mursall** [reply]
Funnily enough I used the "Hello World + KTor"-Template for my project, which still runs on an older version, and upgraded Korge myself.
And those two lines are not present in that template, which is why I was able to build my project without any problems :o

**Reply by meoowe** [reply]
it worked! thank you so much! you are the most helpful person alive.

**Reply by maik-mursall** [reply]
Nice 🎉 
Happy to help :D

**Comment by meoowe** [comment]
Sorry to disturb you yet again but, whenever I package my project using `packageJvmWindowsApp` or any other way, I can't see any sprites or images on anything besides the starting screen. If I use `runJvm` or `runJs` it works.  I have three different scenes: a `TitleScreen`, an `options` scene and a `GameScene`. On the options scene, I can not even see the UI buttons. The   TitleSceen works perfectly fine. On the GameScene I can see the UI buttons. Can

**Reply by maik-mursall** [reply]
Haha don't worry about it :)
I'm here to help :D

I looked at your Game (The repository is public on your GitHub profile :b) and I found your problem:
The images/sprites don't load because you have some typos when accessing some ressources 😅 

E.g: 
`resourcesVfs["img/dog.png"]` should be `resourcesVfs["img/dog.PNG"]` (Notice the upper-case "PNG")

This is some weird behaviour from Korge tho, since it seems to be case insensitive when running using `runJVM` and then becomes case sensitive after packaging 🤔 

And in your case because it couldn't find the "house"-Sprite, it aborted the entire `sceneMain` function 😅

**Reply by maik-mursall** [reply]
IMO Korge should have some better error-handling in this case, since this was pretty tricky to find you 😅

**Reply by maik-mursall** [reply]
One thing that would eliminate this issue is using `KR` instead of strings for accessing your ressources :)
This is an generated object, which holds your ressources. It gets generated during the build process.

E.g
Instead of `resourcesVfs["img/house.png"].readBitmap()` you can just call `KR.img.house.read()`.

The object follows your ressource structure.
If you add any ressource you have to re-build the app for `KR` to get regenerated :)

I hope with that you won't encounter these weird issues anymore :D

**Reply by meoowe** [reply]
Thank you again! I think the upper case .PNG is because it came from an Apple device.
I have never actually used KR so this is all new to me.

**Reply by soywiz** [reply]
It seems that it triggered this: https://github.com/korlibs/korge/issues/599
When using runJvm, maybe it is using resources from the local filesystem that is case insensitive on windows, while packed it is inside a .jar, where it might be case-sensitive.
